kbchat/kbchat_test.go

```diff
@@ -103,82 +102,284 @@ func deleteWorkingDir(workingDir string) error {
 	return os.RemoveAll(workingDir)
 }
 
-var kbc *API
-var config Config
+type testSetupOptions struct {
+	OneOnOnePartner string
+	TeamName        string
+}
 
-func TestMain(m *testing.M) {
-	var err error
-	config, err = readAndParseConfig()
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "Error in reading config: %v\n", err)
-	}
-	dir, err := randomTempDir()
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "Error in generating directory: %v\n", err)
+func testSetup(botName string, options *testSetupOptions) (bot *API, config botConfig, dir string, oneOnOneChannel Channel, teamChannel Channel) {
+	var oneOnOnePartner string
+	var teamName string
+	if options == nil {
+		oneOnOnePartner = "charlie1"
+		teamName = "acme"
+	} else {
+		oneOnOnePartner = options.OneOnOnePartner
+		teamName = options.TeamName
 	}
+
+	config = readAndParseConfig()
+	dir = randomTempDir()
 	kbLocation, err := prepWorkingDir(dir)
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "Error in preparing working directory: %v\n", err)
+		defer deleteWorkingDir(dir)
+		panic(err)
 	}
-	kbc, err = Start(RunOptions{KeybaseLocation: kbLocation, HomeDir: dir, Oneshot: config.Bots.Alice1, StartService: true})
+	bot, err = Start(RunOptions{KeybaseLocation: kbLocation, HomeDir: dir, Oneshot: config.Bots[botName], StartService: true})
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "Error in starting service: %v\n", err)
+		defer testTeardown(bot, dir)
+		panic(err)
 	}
 
-	flag.Parse()
-	code := m.Run()
+	oneOnOneChannel = Channel{
+		Name: fmt.Sprintf("%s,%s", config.Bots[botName].Username, oneOnOnePartner),
+	}
+	teamChannel = Channel{
+		Name:        config.Teams[teamName].Name,
+		Public:      false,
+		MembersType: "team",
+		TopicName:   config.Teams[teamName].TopicName,
+		TopicType:   "chat",
+	}
+
+	return bot, config, dir, oneOnOneChannel, teamChannel
+}
 
-	err = kbc.Shutdown()
+func testTeardown(bot *API, dir string) {
+	err := bot.Shutdown()
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "Error on service shutdown: %v\n", err)
+		panic(err)
 	}
 	err = deleteWorkingDir(dir)
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "Error working directory teardown: %v\n", err)
+		panic(err)
+	}
+}
+
+func getMostRecentMessage(bot *API, channel Channel) Message {
+	messages, err := bot.GetTextMessages(channel, false)
+	if err != nil {
+		panic(err)
 	}
-	os.Exit(code)
+	return messages[0]
+}
+
+func getConvIDForChannel(bot *API, channel Channel) string {
+	messages, err := bot.GetTextMessages(channel, false)
+	if err != nil {
+		panic(err)
+	}
```

Comment:
there's a not uncommon pattern in go...
```go
func panicIfError(err error) {
    if err != nil {
        panic(err)
    }
}
...
messages, err := bot.GetTextMessages(channel, false)
panicIfError(err)
```
that said, in tests, if you're making helper functions that are intended only to be used in tests, i prefer passing in the testing object and using the test helpers like...
```go
func getConvIDForChannel(t *testing.T, bot *API, channel Channel) string {
    messages, err := bot.GetTextMessages(channel, false)
    require.NoError(t, err)
    return messages[0].ConversationID
}
```

Reply:
Alright, I'll go the testing object route